### PR TITLE
Add support for lazy loaded images

### PR DESCRIPTION
**Pull request by Alch-Emi**
The way this patch works is by resolving any `data-src` tags on images in
the same way as normal source tags are resolved.  It is assumed that most
lazy-load libraries will use this tag, and that if this tag is set, then it is a
URL that is in use.

I'm not sure if this is an oversimple solution, but it does seem to resolve lazy loading the the two sites given in the PR that this would close, #16.

### CODE REVIEW

**Comment by snshn** [comment]
Thank you for yet another improvement, Emi.
I wasn't aware of lazy-loading libraries utilizing that attribute... do you think it might be worth checking if the `src` tag is present on this ``, and if not, rename `data-src` into just `src` to make the thing look fine even if JS is disabled or fails to rename that attribute when the page gets rendered?

**Comment by Alch-Emi** [comment]
That shouldn't be too hard!  I'm a bit busy today, but I should have time to work on it at around 18:00-0500

**Comment by snshn** [comment]
Thank you! I'm okay to merge it now, as it gets rid of the main issue, but renaming `data-src` to `src` (or copying `data-src` to `src`) would make it work even when JS is off, which is a very good addition to this enhancement.

**Comment by Alch-Emi** [comment]
I can't make accurate predictions about the length of time projects will take me :p

Lazyloaded images are now *extra* snappy!  Even though an interesting side effect of this is that lazy loaded pages never see the images load, for example on the lazy jq demo, where it becomes stuck at "0 / 10 elements loaded".  We could also work on implementing a few other lazy load features, [like lazy loading bg images][1], but I don't know if that should all go in one PR?

[1]: http://jquery.eisbehr.de/lazy/example_load-background-images

File: src/html.rs [hunk 1]
```diff
@@ -157,36 +157,51 @@ pub fn walk_and_embed_assets(
...
+                            name: QualName::new(None, ns!(), local_name!("src")),
+                            value: Tendril::from_slice(TRANSPARENT_PIXEL),
+                        });
+                    } else if let Some((dataurl, _)) = (&found_datasrc)
+                        .into_iter()
+                        .chain(&found_src) // Give dataurl priority
+                        .map(|attr| &attr.value)
+                        .filter(|src| !src.is_empty()) // Ignore empty srcs
+                        .next()
+                        .and_then(|src| resolve_url(&url, src).ok()) //Make absolute
```
Comment (snshn) [inline src/html.rs]: 👾

File: src/html.rs [hunk 2]
```diff
@@ -157,36 +157,51 @@ pub fn walk_and_embed_assets(
...
+                        if name.eq_ignore_ascii_case("src") {
+                            found_src = Some(attrs_mut.remove(i));
+                        } else if name.eq_ignore_ascii_case("data-src") {
+                            found_datasrc = Some(attrs_mut.remove(i));
+                        } else {
+                            i += 1;
                         }
                     }
+
+                    // If images are disabled, clear both sources
```
Comment (snshn) [inline src/html.rs]: I believe both src and data-src are clear at this point, we just decide to either set src="<transparent_pixel>" if opt_no_images==true, or replace them with dataURL asset strings, but please correct me if I'm wrong

**Comment by snshn** [comment]
I wasn't aware of lazy loading for background images, but I think if it uses data-src we should take care of it as well. Totally up to you if you'd like to implement it here, in a separate branch, or if do it at all. Before your comment I had no idea it was a thing so I'd feel bad asking you to implement it.

I'm sorry if I wasn't specific enough in my comment. After giving it some thought, here's what I think:

|                    | JS enabled                 | JS disabled                                                                       |
|--------------------|----------------------------|-----------------------------------------------------------------------------------|
| img src + data-src | both should become dataURL | src should become a dataURL, data-src should become an absolute URL               |
| img data-src only  | data-src becomes a dataURL | data-src becomes an absolute URL, we copy data-src into src and make it a dataURL |
| div data-src       | data-src becomes a dataURL | data-src becomes an absolute URL (but not a dataURL)                              |

If images are disabled, then all dataURL in that table should be just replaced with the transparent pixel.

I believe it's a good thing that the page's images are already loaded when saved with monolith with this enhancement, but it feels a bit wrong that it changes the behavior of the page. If it implies that data-src is always used for lazy loading and always the same way, then when some JS on that page trips out on having to process the img which already has `src` there, it will technically be monolith's fault since it messed with the code too much while saving it. I wouldn't want it to break whatever lazy loading implementation they have, hence I'm a bit on the fence regarding deviating from how the page operates when it's saved with both JS and images enabled. Please let me know what you think.

**Comment by Alch-Emi** [comment]
In the case that JS is disabled and both `data-src` and `src` in enabled,  leaving `src` in place without replacing it would lead to the page displaying using only very low-res placeholders.  Are we sure that's what we want?  I don't imagine any scripts would be easily tripped up by that swapping out data urls, while JS is off.  I suppose there's the off-chance that someone uses `data-src` on an image to mean something else, but I feel like that seems very unlikely, and more the result of very bad programing practice than anything.

I think I might implement background urls in a new PR.  I think it's likely to involve a fair amount of new code, and might take a while.

**Comment by snshn** [comment]
That's a good point, lemme quickly update the table.

I'm now thinking about your comment regarding how much more snappy the document becomes if we replace the low-res `src` with `data-src` (while getting rid of `data-src`)... it's the top-left cell (**img src + data-src** x **JS enabled**) -- considering what you've just described being a better alternative to the current "both should become dataURL". More so, I think it'd make more sense if we move data-src into src and then convert it to dataURL in all 4 cases for `img`, with `data-src` attributes being removed from those image nodes. What do you think?

**Comment by Alch-Emi** [comment]
Are you suggesting that no matter what, as long as a `data-url` is present on an image, that it always replace the `src` attribute?

**Comment by snshn** [comment]
So far it seems the most reasonable to me, yes. Basically I'm trying to think of cases where the page gets saved with JS either on or off, but then it gets opened in a browser with JS turned off globally, hence if we don't change data-src to just src it'd not show up without JS. Keeping original data-src may cause unnecessary overhead; and converting lo-res images within initial src would also cause an overhead. It sounds like you know more than I do about this whole lazy loading thing, hence I'd like to hear your opinion very much.

**Comment by snshn** [comment]
I skimmed through SingleFile's source (https://github.com/gildas-lormeau/SingleFile/blob/b8922b271422966f6de169188fab4728e00f8c0e/lib/single-file/single-file-core.js#L526) and it looks like it has similar logic, yet only moves data-src into src if the src is unset. There's also some code related to dealing with data-srcset there, but that seems to be outside of the scope of this PR. If we account for JS not being available I think it may be worth to move data-src into src like I mentioned in my previous comment, I can't come up with a better way to treat lazy loaded images. I'll keep data-src in mind and either make --no-images remove it or set to TRANSPARENT_PIXEL to ensure images don't get lazy loaded on imageless pages.

**Comment by snshn** [comment]
Uh-oh, seems like whatever change they did to the formatter in the nightly is now in beta, and it fails all PRs; has nothing to do with your code. I'll fix it in #84, merge into master, and then this branch should be good to go

**Pull request merged**